### DISCUSSION TITLE
Have the ORM handle seamlessly update (eg. delete and replace) of relationships

### DISCUSSION THREAD

**Discussion by scd75** [sqlalchemy/sqlalchemy#6328] [Usage Questions]
**Is your feature request related to a problem? Please describe.**
I find it super easy to update a DB entry through SQLAlchemey's ORM, by doing :
```
my_model.attribute = new_value
db.session.commit()
```
Although that works fine for a model's attribute, the update of relationships throught he same process is a little bit more tricky. In case of a one-to-many relationship, I can do the following:
```
my_model.child_related_objects = new_list
db.session.commit()
```
That works fine if the new objects added in new_list do not break any uniqueness constraints in the Child Table. Yet when I have uniqueness constraints involving foreign key in child table, this operation result in a constaint violation upon commit. I solved it like that:
```
my_model.child_related_objects = []
db.session.flush() # here underlying objects are marked for delete from underlying tables, thanks to my DB cascading setup
my_model.child_related_objects = new_list
db.session.commit()
```
And that works...

**Describe the solution you'd like**
Yet it would be great to have the update (eg. delete and replace) of "Relationship" attributes work in a single command

**Describe alternatives you've considered**
see above

**Have a nice day!**

**Comment by CaselIT** [comment]
Hi,

I suggest reading about cascades in the documentation that seem to be what you seek: https://docs.sqlalchemy.org/en/14/orm/cascades.html?highlight=cascades

**Comment by zzzeek** [comment]
> **Is your feature request related to a problem? Please describe.**
> I find it super easy to update a DB entry through SQLAlchemey's ORM, by doing :
> 
> ```
> my_model.attribute = new_value
> db.session.commit()
> ```
> 
> Although that works fine for a model's attribute, the update of relationships throught he same process is a little bit more tricky. In case of a one-to-many relationship, I can do the following:
> 
> ```
> my_model.child_related_objects = new_list
> db.session.commit()
> ```
> 
> That works fine if the new objects added in new_list do not break any uniqueness constraints in the Child Table. Yet when I have uniqueness constraints involving foreign key in child table

SQLAlchemy does this switch-and-replace operation you mention if you do Python slice assignment:

    my_model.child_related_objects[:] = new_list

this would be the feature you're looking for.  However, if you have the case where some of these records are to be DELETEd, and then replaced with new records that have the same value for the unique constraint, SQLAlchemy does not automatically cover this case.  This is called the "delete before insert" problem and can be reviewed at #2501 ; SQLAlhcemy covers such a case only for primary keys being switched but not other arbitrary columns that would otherwise have a UNIQUE constraint conflict.

In this case, you want to emit a [session.flush](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=session%20flush#sqlalchemy.orm.Session.flush) for the objects to be DELETEd first .

**Comment by scd75** [comment]
Hi,

thank you for your replies. @CaselIT , yes I already have all the cascading setup (ondelete="CASCADE" on child objects etc..)
my problem is more about having to do 2 flushes to first do the delete, then being able to re-insert values which have the same values as some of the previously deleted records (then breacking the UNIQUE constraint).
Anyway, thanks @zzzeek, it is clear. I will keep monitoring the issue #2501 then.

**Comment by CaselIT** [comment]
In postgres ,you can also defer the constraints until the commit. This way an integrity error would not be raised if two rows refer to the same id momentarily.
I would guess something similar is also supported by other databases.
